refactor(extractor/babel)!: sealed, non-fragile babel extractor options

thekip opened the pull request:
Fixes:

#1314 
#1305
#1206
#952

All current problems with babel extractor due to the fact user may have various babel configuration and transformation which potentially may break extraction process (and they do!) 

Briefly: 

1. Extractor should not be affected by any of the babel plugins / preset user may or may not use on the project (hello nextjs with SWC!)
3. Extractor should successfully parse files as long as it is valid ECMA syntax (up to stage3), with exceptions to typescript and flow
4. Extractor should be self-contained, user should not install additional packages to parse a valid ECMAScript code 
6. Extractor should not transform code, only analyze. This brings speed for process and resolve issues when AST after transforming become not parsable by extractor

Supported projects:
- **Typescript / TSX** -> supported out of the box, by matching a file extension
- **JS / JSX** -> supported out of the box
- **Flow** -> supported by enabling `extractorParserOptions.flow: true`
- **Custom** projects with unsupported syntax, stage0 / stage1 proposals, vue, etc -> supported via custom extractor

Custom extractor interface was changed to supply `code` and `sourcemaps`, so simple custom extractor may look like: 

```ts
import babel from "@lingui/cli/api/extractors/babel"

const extractor: ExtractorType = {
  match(filename: string) {
      return filename.endsWith(".custom");
  }

  extract(filename: string, code: string, onMessageExtracted) {
     const {code, sourcemaps} = transformMyCustomFileToJs(filename, code);

     // reuse extractor from cli
     return babel(filename, code, onMessageExtracted, {sourcemaps})
  }
}
```

BREAKING CHANGES:

- @lingui/cli/api/extractors/typescript extractor was removed and should be deleted from lingui.config
- extractorBabelOptions - was removed fron lingui config and no longer supported (todo: write a deperecation warning in configuration validation) and should be removed
- Public interface of ExtractorType is changed


Tasks: 
- [x] Update docs
   - [x] Flow Setup
   - [x] Describe settings
   - [x] How to create a custom extractor
- [x] Add warning in validators for lingui config when extractorBabelOptions passed

andrii-bodnar commented:
@thekip anyway thanks for your contributions, but could you please do it in smaller reviewable parts? Such PRs are extremely hard to review and there is a big risk to miss some bugs

also, please do not push so hard with these major changes, rewriting the library it's something that we don't want at the moment. Please, let's do all of this work in smaller steps

thekip commented:
> but could you please do it in smaller reviewable parts?

No, i could not. This already an atomic change to only one part of functionality.  Please review only this commit https://github.com/lingui/js-lingui/pull/1367/commits/5c162c2dc7b106b4600bac5ba695a470a295aa2a 

The major changes concentrated around one file `@lingui/cli/api/extractors/babel` rest is tests updates or some fixes to make it work.

thekip commented on examples/js/package.json:
```diff
@@ -16,11 +16,11 @@
   "devDependencies": {
-    "@babel/core": "^7.12.3",
-    "@babel/preset-env": "^7.12.1",
+    "@babel/core": "^7.20.12",
+    "@babel/preset-env": "^7.20.2",
```
Babel update was a must to support ES2022 syntax in babel parser

Due to monorepo babel should be updated everywhere  (it brings troubles if different packages has babel versions mismatch)

thekip commented on packages/babel-plugin-extract-messages/test/index.ts:
```diff
@@ -38,15 +37,16 @@ const transformCode = (code: string, filename = 'test-case.js', rootDir=".") =>
       plugins: [
         "@babel/plugin-syntax-jsx",
-        "macros",
+        ["macros", {
+          lingui: { extract: true },
+        },],
```
Instead of fighting with `process.NODE_ENV == production` just explicitly pass option to the babel macros. When this option is set macro would not strip non essentials props.

thekip commented on packages/cli/src/api/extractors/babel.ts:
```diff
@@ -1,41 +1,62 @@
 const babelRe = new RegExp(
-	"\\.(" + [...DEFAULT_EXTENSIONS, ".ts", ".tsx"].map(ext => ext.slice(1)).join("|") + ")$",
-	"i"
+  "\\.(" + [...DEFAULT_EXTENSIONS, ".ts", ".mts", ".cts", ".tsx"].map(ext => ext.slice(1)).join("|") + ")$",
+  "i"
```
We should take care of new native ESM file extensions

thekip commented on packages/cli/src/api/extractors/babel.ts:
```diff
@@ -1,41 +1,62 @@
-    const { plugins = [], ...babelOptions } = _babelOptions
-    const frameworkOptions: BabelOptions = {}
+  async extract(filename, code, onMessageExtracted, options = {}) {
+    const parserPlugins: ParserPlugin[] = [
+      // https://babeljs.io/docs/en/babel-parser#latest-ecmascript-features
+      ["decorators", {
+        decoratorsBeforeExport: options?.parserOptions?.decoratorsBeforeExport || true
+      }]
+    ]
```
By default, babel parser can understand syntax up to es2022, no additional parser plugins needed. 
We add here support for decorators because they are extremely popular in TS projects.

thekip commented on packages/cli/src/api/extractors/babel.ts:
```diff
@@ -1,41 +1,62 @@
+      // don't generate code
+      code: false,
+
+      babelrc: false,
+      configFile: false,
```
don't allow user's babel.config to broke extraction process

thekip commented on packages/cli/src/api/extractors/index.ts:
```diff
@@ -1,11 +1,11 @@
-export type BabelOptions = {
-  plugins?: Array<string>
-  presets?: Array<string>
+export type ParserOptions = {
+  decoratorsBeforeExport?: boolean
+  flow?: boolean,
```
The general idea is to hide implementation details of extractor. User should not know that babel parser is used. Tomorrow we can switch parser to SWC for example without breaking anything

thekip commented on packages/cli/src/api/extractors/typescript.ts:
```diff
@@ -1,63 +1,13 @@
 const extractor: ExtractorType = {
   match(filename) {
-    return typescriptRe.test(filename)
+    throw new Error("Typescript extractor was removed. " +
+      "Lingui CLI can parse typescript out of the box. " +
+      "Please remove it from your lingui.config.js");
```
That was deprecated long time ago, now it's completely unused. People still used it because of burden in resulting babel configuration, so for some configurations this worked better.  Now it's completely unneeded.

thekip commented on packages/cli/src/api/fixtures/collect-typescript-jsx/macro.tsx:
```diff
@@ -0,0 +1,25 @@
+import { t, defineMessage, plural, Trans } from "@lingui/macro"
+
+// JS Macro usages
+const msg = t`Message`
+
+const withDescription = defineMessage({
+  message: 'Description',
+  comment: "description"
+})
+
+const withTId = t({
+    id: "ID Some",
+    message: "Message with id some"
+  })
+
+// JSX Macro usages
+;<Trans>Hi, my name is {name}</Trans>
+;<Trans context="Context1">Some message</Trans>
+;<span title={t`Title`} />
+;<span
+  title={plural(count, {
+    one: "# book",
+    other: "# books"
+  })}
+/>
```
There was no test with macro in CLI package at all, I added one. Here we just generally check that macro is applied. in extractor More macro extraction cases in babel-extractor-plugin

thekip commented on packages/cli/src/lingui-extract.ts:
```diff
@@ -102,36 +102,13 @@ if (require.main === module) {
       "Convert from previous format of message catalogs"
     )
     .option("--watch", "Enables Watch Mode")
-    // Obsolete options
-    .option(
-      "--babelOptions",
-      "Babel options passed to transform/extract plugins"
-    )
-    .option("--format <format>", "Format of message catalogs")
```
it's time to delete a deprecated option from v2

thekip commented on packages/macro/src/index.ts:
```diff
@@ -51,15 +58,17 @@ function macro({ references, state, babel }: MacroParams) {
   jsNodes.filter(isRootPath(jsNodes)).forEach((path) => {
     if (alreadyVisited(path)) return
-    const macro = new MacroJS(babel, { i18nImportName })
+    const macro = new MacroJS(babel, { i18nImportName, stripNonEssentialProps })
```
I also encountered that only JSX macro stripping non essential props.  JS macro doesn't have such logic, we need to fix it next iterations

thekip commented on packages/macro/test/fixtures/js-t-var/js-t-var.expected.js:
```diff
@@ -1,20 +1,15 @@
-
-    /*i18n*/
-    _core.i18n._("This is bar {bar}", {
+    /*i18n*/_core.i18n._("This is bar {bar}", {
```
The test fixtures in macro/test was updated because of newer babel/generator. Looks like newer version generates slightly different output in terms of whitespaces. This not affect logic anyhow, only formatting. 

So next you will see many changes related to this fact.

thekip commented on packages/cli/package.json:
```diff
@@ -83,10 +82,10 @@
   },
   "peerDependencies": {
-    "@babel/core": "^7.0.0",
     "babel-plugin-macros": "2 || 3",
```
this should be moved to dependencies, instead of peer, as for now we are not relaying on the users environment but providing all needed dependencies by ourselves.

thekip commented on packages/cli/src/lingui-extract-template.ts:
```diff
@@ -24,11 +22,6 @@ export default async function command(
     process.env.BABEL_ENV = "development"
   }
 
-  // We need macros to keep imports, so extract-messages plugin know what componets
-  // to collect. Users usually use both BABEN_ENV and NODE_ENV, so it's probably
-  // safer to introduce a new env variable. LINGUI_EXTRACT=1 during `lingui extract`
-  process.env.LINGUI_EXTRACT = "1"
-
```
That flag was used to tell babel-macro plugin to not delete macro imports. 
That actually don't need. Because extractor analyze code AFTER macro transformation and lokiing for `i18n._` and `/*i18n*/` statements, not a macro calls.

thekip commented on packages/cli/src/api/extractors/babel.ts:
```diff
@@ -1,41 +1,62 @@
+    const parserPlugins: ParserPlugin[] = [
+      // https://babeljs.io/docs/en/babel-parser#latest-ecmascript-features
+      ["decorators", {
+        decoratorsBeforeExport: options?.parserOptions?.decoratorsBeforeExport || true
```
Support stage3 decorators as well as legacy syntax

Martin005 commented:
@thekip I went through the changes and they look great! After the #1358 is merged, could you please finish the tasks you mentioned at the end of your PR description?

thekip commented:
@Martin005 waiting for [this one](https://github.com/lingui/js-lingui/pull/1415) merging before can continue to work on PR

thekip commented:
@andrii-bodnar may be it's time to dive into the context of changes for the library which you took responsibility?

thekip commented:
Added docs in a separate commit. 

> Add warning in validators for lingui config when extractorBabelOptions passed

This is waiting for https://github.com/lingui/js-lingui/pull/1415

andrii-bodnar commented:
> Added docs in a separate commit

@thekip thank you!

andrii-bodnar commented on website/docs/guides/advanced-custom-extractor.md:
```diff
@@ -0,0 +1,36 @@
+# Advanced: Custom Extractor
```
```suggestion
# Custom Extractor
```

I'd suggest removing this prefix since such a heading format is something unusual for this site

\+ rename file

thekip replied:
That's exactly how i wanted to write this. There will be more in the future. If you want you can create a separate group in the sidebar for that. But this is totally fine to have "advanced" https://nextjs.org/docs/advanced-features/compiler

andrii-bodnar replied:
Let's create a separate "Advanced" category. This could be a sub-category inside the "Guides" category.

andrii-bodnar commented on website/docs/guides/advanced-custom-extractor.md:
```diff
@@ -0,0 +1,36 @@
+
+If you not project not playing well with Lingui's Extractor you can write your custom extractor implementation.
+
+That's might be the case if you use some experimental features (stage0 - stage2) or frameworks with custom syntax such as Vue or Svelte.
```
```suggestion
That might be the case if you use some experimental features (stage0 - stage2) or frameworks with custom syntax such as Vue.js or Svelte.
```

andrii-bodnar commented on website/docs/guides/advanced-custom-extractor.md:
````diff
@@ -0,0 +1,36 @@
+```
+
+:::caution Important
+If you use typescript for creating your extractor, you should also use `ts` extension for your lingui config file.
````
```suggestion
If you use TypeScript to create your extractor, you should use the `ts` extension for your Lingui configuration file.
```

andrii-bodnar commented on website/docs/guides/flow.md:
```diff
@@ -0,0 +1,11 @@
+# Flow
+
+Lingui already supports [Flow](https://flow.org/) by default. All you need to do is enable it in your config file.
```
Can we add more details on how the new flag works in this case? The article seems a bit poor

thekip replied:
When you enable flag, extractor will understand flow. That's it. So simple.

andrii-bodnar commented on website/docs/releases/migration-4.md:
````diff
@@ -0,0 +1,69 @@
+
+```diff title="lingui.config.js"
+module.exports = {
+  - extractorBabelOptions: { [...] }
````
the diffs are not highlighted, please check the preview - https://js-lingui-git-fork-thekip-refactor-extractor-babel-lingui.vercel.app/releases/migration-4

thekip replied:
done

andrii-bodnar commented on website/docs/releases/migration-4.md:
````diff
@@ -0,0 +1,69 @@
+}
+```
+
+#### Flow Syntax Supported in the Extractor with the flag
````
```suggestion
#### Flow Syntax supported in the Extractor with the flag
```

andrii-bodnar commented on website/docs/releases/migration-4.md:
````diff
@@ -0,0 +1,69 @@
+```
+
+### `@lingui/cli/api/extractors/typescript` was deleted
+Extractor supports typescript out of the box. Please delete it from your configuration file.
````
```suggestion
Extractor supports TypeScript out of the box. Please delete it from your configuration file.
```

andrii-bodnar commented on website/docs/releases/migration-4.md:
```diff
@@ -0,0 +1,69 @@
+
+#### Flow Syntax Supported in the Extractor with the flag
+
+If your project uses Flow you, you need explicitly enable support in the extractor:
```
```suggestion
If your project uses Flow, you need explicitly enable support in the extractor:
```

thekip commented:
Guys, i added 2 more commits. First - adds deprecation warning for `extractBabelOptions` 

Second, removes `extractBabelOptions` from monorepos documentation. I deleted outdated part, but this article is definitely require more work, if you have ideas how that page could be improved, let's do that in next iterations

Martin005 commented on packages/conf/src/__snapshots__/index.test.ts.snap:
```diff
@@ -1,5 +1,18 @@
+    
+    Please remove it from your config file. 
+
+    You can find more information here: https://lingui.dev/releases/migration-4.md
```
```suggestion
    You can find more information here: https://lingui.dev/releases/migration-4
```

Martin005 commented on website/docs/releases/migration-4.md:
```diff
@@ -0,0 +1,67 @@
+
+### Extractor configuration changes
+
+The big change in the v4 changes in extractor internals. Now it is less fragile, and not depends on the host project settings.
```
```suggestion
The big change in the v4 is in extractor internals. Now it is less fragile, and doesn't depend on the host project settings.
```

Martin005 commented on website/docs/releases/migration-4.md:
```diff
@@ -0,0 +1,67 @@
+
+#### Flow Syntax supported in the Extractor with the flag
+
+If your project uses Flow, you need explicitly enable support in the extractor:
```
```suggestion
If your project uses Flow, you need to explicitly enable support in the extractor:
```

Martin005 commented on website/docs/releases/migration-4.md:
````diff
@@ -0,0 +1,67 @@
+If your extract command looks like:
+
+```bash
+ NODE_ENV=development lingui-extract
````
```suggestion
NODE_ENV=development lingui-extract
```

Martin005 commented on website/docs/releases/migration-4.md:
````diff
@@ -0,0 +1,67 @@
+Now you can safely change it to just:
+
+```bash
+ lingui-extract
````
```suggestion
lingui-extract
```

Martin005 commented on website/docs/releases/migration-4.md:
````diff
@@ -0,0 +1,67 @@
+}
+```
+
+Read more about custom extractor on the [Custom Extractor Page](/guides/custom-extractor)
````
```suggestion
Read more about custom extractor on the [Advanced: Custom Extractor](/guides/custom-extractor.md) page.
```

thekip replied:
with `.md` at the end?

Martin005 replied:
@thekip Oh, sorry, `.md` is not mandatory if you don't use anchor tags to link to a specific section of a page. I have updated my suggestion.

Martin005 commented on website/docs/ref/conf.md:
```diff
@@ -432,20 +439,13 @@ The difference between [`fallbackLocales`](#fallbacklocales) and `sourceLocale`
-
-- match: regex to a filename extension, should return `true`|`false`
-- extract: is the responsible for transforming the code and using @lingui/babel-plugin-extract-messages
+Visit [Advanced: Custom Extractor](/guides/custom-extractor.md) to learn more how to create custom extractor.
```
```suggestion
Visit [Advanced: Custom Extractor](/guides/custom-extractor.md) to learn how to create custom extractor.
```

Martin005 commented on website/docs/guides/custom-extractor.md:
```diff
@@ -0,0 +1,36 @@
+# Custom Extractor
+
+If your project is not playing well with Lingui's Extractor you can write your custom extractor implementation.
```
```suggestion
If your project is not working well with Lingui's Extractor, you can write your custom extractor implementation.
```

Martin005 commented on website/docs/guides/custom-extractor.md:
```diff
@@ -0,0 +1,36 @@
+      return filename.endsWith(".custom");
+  },
+  extract(filename: string, code: string, onMessageExtracted) {
+    // transform to plain JS + Sourcemaps
+     const {code, sourcemaps} = transformMyCustomFileToJs(filename, code);
+
+     // reuse extractor from cli
+     return babel(filename, code, onMessageExtracted, {sourcemaps})
```
```suggestion
    // transform to plain JS + Sourcemaps
    const {code, sourcemaps} = transformMyCustomFileToJs(filename, code);

    // reuse extractor from cli
    return babel(filename, code, onMessageExtracted, {sourcemaps})
```

Martin005 commented on website/docs/guides/custom-extractor.md:
````diff
@@ -0,0 +1,36 @@
+Then in your `lingui.config.ts`:
+
+```ts title="lingui.config.ts"
+import {extractor} from './my-custom-extractor.ts'
````
```suggestion
import { extractor } from './my-custom-extractor.ts'
```

Martin005 commented on website/docs/ref/conf.md:
```diff
@@ -207,17 +207,24 @@ For example, setting [`compileNamespace`](#compilenamespace) to `window.i18n` cr
+   * Use for Stage 3 decorators syntax
+   */
+  decoratorsBeforeExport?: boolean
+  /**
+   * Enable if you use flow. Default false
+   * This will apply Flow syntax to files with .js, cjs, .mjs extension.
+   */
```
```suggestion
  /**
   * default false
   * Enable if you use flow. This will apply Flow syntax to files with .js, cjs, .mjs extension.
   */
```

Martin005 commented on website/docs/ref/conf.md:
```diff
@@ -207,17 +207,24 @@ For example, setting [`compileNamespace`](#compilenamespace) to `window.i18n` cr
+"extractorParserOptions": {
+  /**
+   * default true
+   * Use for Stage 3 decorators syntax
```
```suggestion
   * Use for Stage 3 decorators syntax.
```

thekip commented:
done
